src/internal_modules/roc_core/pairing_heap.h

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright (c) 2015 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_core/pairing_heap.h
+//! @brief Intrusive pairing heap.
+
+#ifndef ROC_CORE_PAIRING_HEAP_H_
+#define ROC_CORE_PAIRING_HEAP_H_
+
+#include "roc_core/noncopyable.h"
+#include "roc_core/ownership_policy.h"
+#include "roc_core/pairing_heap_node.h"
+#include "roc_core/panic.h"
+#include "roc_core/stddefs.h"
+
+namespace roc {
+namespace core {
+
+//! Intrusive pairing heap.
+//!
+//! Does not perform allocations.
+//! Provides O(1) size check, membership check, insertion, and removal.
+//!
+//! @tparam T defines object type, it should inherit PairingHeapNode.
+//!
+//! @tparam OwnershipPolicy defines ownership policy which is used to acquire an
+//! element ownership when it's added to the pairing heap and release ownership when it's
+//! removed from the pairing heap.
+template <class T, template <class TT> class OwnershipPolicy = RefCountedOwnership>
+class PairingHeap : public NonCopyable<> {
```

Comment:
The interface of this class is quite low-level as it reflects the underlying algorithm.

I suggest to use different approach: provide much more narrow interface which is actually needed for the user.

What we need in roc_ctl is a classic priority queue with just three operations:

* push (insert element according to its priority)
* top (look up element with highest priority)
* pop (remove element with highest priority)

So I suggest to rename this class to PriorityQueue<T> and hide all internals behind these high-level operations.

Benefits:

* interface will be much simpler to understand and use, and there will be less bugs in usage
* interface will depend on algorithm properties, but not on specific algorithm; we will state that it has cheap push and more expensive pop, but we won't attach caller to specific data structure and can change it if needed

To achieve this, we'll need to tell PriorityQueue how it can get priority of T. We can use the same approach as we use in [hashmap](https://github.com/roc-streaming/roc-toolkit/blob/develop/src/internal_modules/roc_core/hashmap.h): T will have to implement necessary methods, for example:

```cpp
   // get object priority
   Prio prio() const;

   // compare two priorities (return -1, 0, or 1)
   static int prio_cmp(Prio prio1, Prio prio2);
```

(See also comments in Hashmap).

I think it would be pretty convenient, since this is intrusive data structure, and we anyway place special requirements on elements (inherit from node class).

What do you think?

Reply:
I will try! 😅

questions:
>     * pop (remove element with highest priority)
Are we assuming we never remove / change deadline of elements other than the top one? Since there is `ControlTaskQueue::cancel_task_` function, I wasn't too sure.

for benchmarks
my contention test freezes here:
<img width="553" alt="contention" src="https://github.com/roc-streaming/roc-toolkit/assets/25291631/fa6a2472-4dbb-41ae-978b-a0dff7d74e5d">
is it just my side?

Reply:
Good questions,

> Are we assuming we never remove / change deadline of elements other than the top one? Since there is ControlTaskQueue::cancel_task_ function, I wasn't too sure.

My bad, we indeed need to remove tasks. The interface then can be:

* `void push(T&)`
* `T* top()`
* `void pop()`
* `bool contains(T&)`
* `void remove(T&)`

BTW, it would be better to use naming similar to List and Hashmap:

* `void insert(T&)`
* `T* front()`
* `void pop_front()`
* `bool contains(T&)`
* `void remove(T&)`

> If this returns 0 I'm assuming priority is equal.

Yep, like in strcmp() and memcmp(). Another option is to make it `static bool prio_less(Prio prio1, Prio prio2)` and return true if `prio1 < prio2`. Choose whatever you think fits better.

> Do we need to make sure which task goes first? or can we do either one.

I think it would be useful to have this guarantee, because it will prevent some easy to make bugs in user code.

Imagine there is a code that calculates next deadline for some event (e.g. periodic timer), and then runs some logic that schedules a few tasks for that deadline, like:

```
next_deadline_ = ...;

...

if (need_x) {
   schedule_at(next_deadline_, task_x);

   ...

   if (also_need_y) {
      schedule_at(next_deadline_, task_y);
   }
}
```

It would be easy to forget that order of task_x and task_y is not guaranteed, especially given that it's guaranteed for regular schedule (without deadline).

> However I don't think we can guarantee this using a heap because of how heap merge works.

But how do you handle it currently? I guess it works somehow because tests pass on this PR?

Also, if needed, we could upgrade PairingHeap (PriorityQueue) so that each node will contain a list of nodes with the same priority. When you add node, and another node with same prio already exists, it will be added to that list. And the list will maintain insertion order.

> my contention test freezes here:

This indicates a bug in benchmarks I think. They work on my machine and a few others where I tested them. (BTW they take quite a lot of time).

It would be great if you can debug the reason of the freeze because I can't reproduce it. Anyway it should be done separately from this PR I think.